core/amber/src/main/scala/edu/uci/ics/texera/web/StateStore.scala

```diff
@@ -0,0 +1,52 @@
+package edu.uci.ics.texera.web
+
+import edu.uci.ics.texera.Utils.withLock
+import edu.uci.ics.texera.web.model.websocket.event.TexeraWebSocketEvent
+import io.reactivex.rxjava3.core.{Observable, Observer, Single}
+import io.reactivex.rxjava3.disposables.Disposable
+import io.reactivex.rxjava3.subjects.{BehaviorSubject, PublishSubject}
+
+import java.util.concurrent.locks.ReentrantLock
+import scala.collection.mutable
+import java.util
+
+class StateStore[T](defaultState: T) {
+
+  private val stateSubject = BehaviorSubject.createDefault(defaultState)
+  private val serializedSubject = stateSubject.toSerialized
+  private implicit val lock: ReentrantLock = new ReentrantLock()
+  private val diffHandlers = new mutable.ArrayBuffer[(T, T) => Iterable[TexeraWebSocketEvent]]
+  private val diffSubject = serializedSubject
+    .startWith(Single.just(defaultState))
+    .buffer(2, 1)
+    .map[Iterable[TexeraWebSocketEvent]] { states: util.List[T] =>
```

Comment:
Let's add a filter here to filter out the case where `prevState == currentState`, just pointer comparison is fine I think. 
This is just for the case where the first subscriber doesn't need to compare two default states (the default from behavior subject, and from startWith)

Reply:
done